runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/DistributionData.java

```diff
@@ -17,48 +17,283 @@
+      this.percentileExtractors = createPercentileExtractors(percentiles);
+    } else {
+      this.sketch = Optional.empty();
+      this.percentileExtractors = ImmutableMap.of();
+    }
   }
 
   public static DistributionData singleton(long value) {
-    return create(value, 1, value, value);
+    final DistributionData distributionData = empty();
+    distributionData.update(value);
+    return distributionData;
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////////////////////////////
+  // Getters
+
+  public long sum() {
+    return sum;
+  }
+
+  public long count() {
+    return count;
+  }
+
+  public long min() {
+    return min;
+  }
+
+  public long max() {
+    return max;
+  }
+
+  public Map<Double, Double> percentiles() {
+    computePercentiles();
+    return percentileResults;
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Updates the distribution with a value. For percentiles, only add the value to the sketch.
+   * Percentile will be computed prior to calling {@link DistributionCell#getCumulative()} or in
+   * {@link #extractResult()}.
+   *
+   * @param value value to update the distribution with.
+   */
+  public void update(long value) {
+    ++count;
+    min = Math.min(min, value);
+    max = Math.max(max, value);
+    sum += value;
+    sketch.ifPresent(currSketch -> currSketch.update(value));
   }
 
-  public DistributionData combine(DistributionData value) {
-    return create(
-        sum() + value.sum(),
-        count() + value.count(),
-        Math.min(value.min(), min()),
-        Math.max(value.max(), max()));
+  /** Merges two distributions. */
+  public DistributionData combine(DistributionData other) {
+    if (sketch.isPresent()
+        && other.sketch.isPresent()
+        && sketch.get().getN() > 0
+        && other.sketch.get().getN() > 0) {
+      final DoublesUnion union = new DoublesUnionBuilder().build();
+      union.update(sketch.get());
+      union.update(other.sketch.get());
+      sketch = Optional.of(union.getResult());
+    } else if (other.sketch.isPresent() && other.sketch.get().getN() > 0) {
+      sketch = other.sketch;
+    }
+    sum += other.sum;
+    count += other.count;
+    max = Math.max(max, other.max);
+    min = Math.min(min, other.min);
+    return this;
   }
 
+  public DistributionData reset() {
+    this.sum = 0L;
+    this.count = 0L;
+    this.min = Long.MAX_VALUE;
+    this.max = Long.MIN_VALUE;
+    if (!this.percentiles.isEmpty()) {
+      final DoublesSketchBuilder doublesSketchBuilder = new DoublesSketchBuilder();
+      this.sketch = Optional.of(doublesSketchBuilder.setK(SKETCH_SUMMARY_SIZE).build());
+    } else {
+      this.sketch = Optional.empty();
+    }
+    this.percentileResults = ImmutableMap.of();
+    return this;
+  }
+
+  /** Gets the value percentiles as a map. */
+  public void computePercentiles() {
```

Comment:
Seems we should return the percentile results map directly here instead of setting a member var. Mark it private?

Reply:
Changed. It was marked public so that it could be called from `DistributionCell`. SImplified it.